core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -1447,8 +1451,9 @@ private[spark] object MapOutputTracker extends Logging {
     // TODO: SPARK-35036: Instead of reading map blocks in case of AQE with Push based shuffle,
     // TODO: improve push based shuffle to read partial merged blocks satisfying the start/end
     // TODO: map indexes
-    if (mergeStatuses.exists(_.nonEmpty) && startMapIndex == 0
-      && endMapIndex == mapStatuses.length) {
+    if (mergeStatuses.exists(_.nonEmpty) && mergeStatuses.exists(_.exists(_ != null))
```

Comment:
does this check still makes sense `mergeStatuses.exists(_.nonEmpty)` as it will always be `Some(Array())` when push-based shuffle flag is set to true only will be `None` in the case when push-based shuffle flag is set to false

Reply:
Agree, `mergeStatuses.exists(_.exists(_ != null)` should be enough now.

Reply:
Ideally we can avoid doing this costly check by checking `numAvailableMergeResults` and setting this `mergeStatuses` to `None` before passing to `convertMapStatuses`. But may be we can do that as a follow-up as a separate PR .

Reply:
@zhouyejoe Do we still need this check `mergeStatuses.exists(_.nonEmpty)`?

Reply:
Thanks. I do have updated the PR to exclude this check.